llama_index/indices/multi_modal/base.py

```diff
@@ -0,0 +1,331 @@
+import logging
+from typing import Any, List, Optional, Sequence
+
+from llama_index.data_structs.data_structs import MultiModelIndexDict
+from llama_index.embeddings.mutli_modal_base import MultiModalEmbedding
+from llama_index.embeddings.utils import EmbedType, resolve_embed_model
+from llama_index.indices.base_retriever import BaseRetriever
+from llama_index.indices.service_context import ServiceContext
+from llama_index.indices.utils import (
+    async_embed_image_nodes,
+    async_embed_nodes,
+    embed_image_nodes,
+    embed_nodes,
+)
+from llama_index.indices.vector_store.base import VectorStoreIndex
+from llama_index.schema import BaseNode, ImageNode, IndexNode
+from llama_index.storage.storage_context import StorageContext
+from llama_index.vector_stores.simple import DEFAULT_VECTOR_STORE, SimpleVectorStore
+from llama_index.vector_stores.types import VectorStore
+
+logger = logging.getLogger(__name__)
+
+
+class MultiModalVectorStoreIndex(VectorStoreIndex):
+    """Multi-Modal Vector Store Index.
+
+    Args:
+        use_async (bool): Whether to use asynchronous calls. Defaults to False.
+        show_progress (bool): Whether to show tqdm progress bars. Defaults to False.
+        store_nodes_override (bool): set to True to always store Node objects in index
+            store and document store even if vector store keeps text. Defaults to False
+    """
+
+    image_namespace = "image"
+    index_struct_cls = MultiModelIndexDict
+
+    def __init__(
+        self,
+        nodes: Optional[Sequence[BaseNode]] = None,
+        index_struct: Optional[MultiModelIndexDict] = None,
+        service_context: Optional[ServiceContext] = None,
+        storage_context: Optional[StorageContext] = None,
+        use_async: bool = False,
+        store_nodes_override: bool = False,
+        show_progress: bool = False,
+        # Image-related kwargs
+        image_vector_store: Optional[VectorStore] = None,
+        image_embed_model: EmbedType = "clip",
+        **kwargs: Any,
+    ) -> None:
+        """Initialize params."""
+        image_embed_model = resolve_embed_model(image_embed_model)
+        assert isinstance(image_embed_model, MultiModalEmbedding)
+        self._image_embed_model = image_embed_model
+
+        storage_context = storage_context or StorageContext.from_defaults()
+
+        if image_vector_store is not None:
+            storage_context.add_vector_store(image_vector_store, self.image_namespace)
+
+        if self.image_namespace not in storage_context.vector_stores:
+            storage_context.add_vector_store(SimpleVectorStore(), self.image_namespace)
+
+        super().__init__(
+            nodes=nodes,
+            index_struct=index_struct,
+            service_context=service_context,
+            storage_context=storage_context,
+            show_progress=show_progress,
+            use_async=use_async,
+            # force to true, since vector dbs don't store images
+            store_nodes_override=True,
```

Comment:
Note a small "gotcha" here -- vector dbs don't (currently) store images, so we force this to true